apps/reaper/lib/reaper/xml/xml_stream/sax_handler.ex

```diff
@@ -0,0 +1,159 @@
+defmodule XMLStream.SaxHandler do
+  @moduledoc false
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys [:tag_path, :emitter]
+    defstruct stack: [], accumulate: false, tag_stack: [], tag_path: [], emitter: nil
+
+    def new(fields \\ %{}) do
+      struct!(__MODULE__, fields)
+    end
+
+    def update(%__MODULE__{}=state, fields) do
+      struct!(state, fields)
+    end
+
+    @spec pop(%State{}, :stack | :tag_stack | :tag_path) :: {any(), %State{}}
+    def pop(%__MODULE__{}=state, key) when key in [:stack, :tag_stack, :tag_path] do
+      [item | rest] = Map.fetch!(state, key)
+
+      {item, State.update(state, [{key, rest}])}
+    end
+
+    def pop_current_tag(%__MODULE__{}=state, tag_name) do
+      {{^tag_name, attributes, content}, state} = State.pop(state, :stack)
+
+      {{tag_name, attributes, Enum.reverse(content)}, state}
+    end
+
+    def at_path?(%__MODULE__{tag_stack: tag_stack, tag_path: tag_path}, tag_name) do
+      [tag_name | tag_stack] == tag_path
+    end
+
+    def ascend(%__MODULE__{}=state) do
+      State.pop(state, :tag_stack)
+    end
+
+    def descend(%__MODULE__{tag_stack: tag_stack}=state, tag_name) do
+      State.update(state, tag_stack: [tag_name | tag_stack])
+    end
+  end
+
+  @behaviour Saxy.Handler
+  @file_chunk_size 40_000
+
+  ######################
+  ## Client Functions ##
+  ######################
+  def start_stream(filepath, tag_path_string, callback) when is_function(callback, 1) do
+    tag_path = parse_selector(tag_path_string)
+
+    filepath
+    |> File.stream!([], @file_chunk_size)
+    |> handle_ufeff()
+    |> Saxy.parse_stream(
+      __MODULE__,
+      __MODULE__.State.new(tag_path: tag_path, emitter: callback)
+    )
+  end
+
+  ###############
+  ## Callbacks ##
+  ###############
+  @impl true
+  def handle_event(:start_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  @impl true
+  def handle_event(:end_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  @impl true
+  def handle_event(:start_element, {tag_name, attributes}, %State{}=state) do
+    if State.at_path?(state, tag_name) or state.accumulate do
+      tag = {tag_name, attributes, []}
+
+      state
+      |> State.descend(tag_name)
+      |> State.update(stack: [tag | state.stack], accumulate: true)
+      |> ok()
+    else
+      state
+      |> State.descend(tag_name)
+      |> ok()
+    end
+  end
+
+  @impl true
+  def handle_event(:characters, chars, %State{stack: stack, accumulate: true} = state) do
+      [{tag_name, attributes, content} | stack] = stack
+
+      current = {tag_name, attributes, [chars | content]}
+
+      state
+      |> State.update(stack: [current | stack])
+      |> ok()
```

Comment:
Didn't look to see if the formatter complained, but the indentation is off here.

Reply:
I didn't run the formatter yet, to more clearly show code changes vs formatting changes

Reply:
When I ran it the formatter didn't do anything. Do you prefer the follow up lines to be indented here?

Reply:
I'm just referring to the entire function block being indented 4 spaces instead of the usual 2.

Reply:
I would have expected the formatter to catch that.